rust/core-lib/src/plugins/manager.rs

```diff
@@ -346,41 +382,58 @@ impl <W: Write + Send + 'static>PluginManager<W> {
     }
 }
 
-/// Wrapper around a `Arc<Mutex<PluginManager<W>>>`.
-pub struct PluginManagerRef<W: Write>(Arc<Mutex<PluginManager<W>>>);
+/// Wrapper around an `Arc<Mutex<PluginManager<W>>>`.
+///
+/// `PluginManagerRef` can also store closures to be executed on the next pass
+/// of the runloop; the second two fields support this functionality.
+pub struct PluginManagerRef<W: Write>(
+    Arc<Mutex<PluginManager<W>>>,
+    Arc<Mutex<Vec<Box<Fn(PluginManagerRef<W>) + 'static + Send>>>>,
+    Arc<AtomicBool>);
 
 impl<W: Write> Clone for PluginManagerRef<W> {
     fn clone(&self) -> Self {
-        PluginManagerRef(self.0.clone())
+        PluginManagerRef(self.0.clone(), self.1.clone(), self.2.clone())
     }
 }
 
 /// Wrapper around a `Weak<Mutex<PluginManager<W>>>`
-pub struct WeakPluginManagerRef<W: Write>(Weak<Mutex<PluginManager<W>>>);
+pub struct WeakPluginManagerRef<W: Write>(
+    Weak<Mutex<PluginManager<W>>>,
+    Weak<Mutex<Vec<Box<Fn(PluginManagerRef<W>) + 'static + Send>>>>,
```

Comment:
The reason these live outside of the `PluginManager` is because the public interfaces exist on `PluginManagerRef`.

The reason it isn't just one `<Arc<Mutex<_>>` is that we want to check for work on each idle call, and I wanted to avoid unnecessary locking, hence using the `AtomicBool` to store whether or not there's work.

But I do agree that these types look hairier than they need to. At the very least I should be able to combine the second two fields into some sort of `AtomicVec` type or something.